ArmPkg/Drivers/CpuDxe/PageTableMemoryAllocation.c

```diff
@@ -0,0 +1,202 @@
+/** @file PageTableAlloc.c
+
+Logic facilitating the pre-allocation of page table memory.
+
+Copyright (C) Microsoft Corporation. All rights reserved.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <Uefi.h>
+#include <Library/BaseLib.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Protocol/ArmPageTableMemoryAllocation.h>
+
+#pragma pack(1)
+
+typedef struct {
+  UINT32        Signature;
+  UINTN         Offset;
+  UINTN         FreePages;
+  LIST_ENTRY    NextPool;
+} PAGE_TABLE_POOL;
+
+#pragma pack()
+
+#define MIN_PAGES_AVAILABLE        5
+#define PAGE_TABLE_POOL_SIGNATURE  SIGNATURE_32 ('P','T','P','L')
+
+UINTN       TotalFreePages     = 0;
+BOOLEAN     mPageTablePoolLock = FALSE;
+LIST_ENTRY  mPageTablePoolList = INITIALIZE_LIST_HEAD_VARIABLE (mPageTablePoolList);
+
+/**
+  Allocates pages for page table memory and adds them to the pool list.
+
+  @param[in] PoolPages The number of pages to allocate. The actual number of pages
+                       reserved will be at least 512 - the number of pages required to describe a
+                       fully split 2MB region.
+
+  @retval EFI_SUCCESS             More pages successfully added to the pool list
+  @retval EFI_INVALID_PARAMETER   PoolPages is zero
+  @retval EFI_ABORTED             Software lock is held by another call to this function
+  @retval EFI_OUT_OF_RESOURCES    Unable to allocate pages
+**/
+EFI_STATUS
+GetMorePages (
+  IN  UINTN  PoolPages
+  )
+{
+  PAGE_TABLE_POOL  *Buffer;
+
+  if (PoolPages == 0) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  PoolPages++;
+  PoolPages = ALIGN_VALUE (PoolPages, EFI_SIZE_TO_PAGES (SIZE_2MB)); // Add one page for the header
+  Buffer    = (PAGE_TABLE_POOL *)AllocateAlignedPages (PoolPages, BASE_2MB);
+  if (Buffer == NULL) {
+    DEBUG ((DEBUG_ERROR, "ERROR: Out of aligned pages\n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  // Reserve one page for pool header.
+  Buffer->FreePages = PoolPages - 1;
+  Buffer->Offset    = EFI_PAGE_SIZE;
+  Buffer->Signature = PAGE_TABLE_POOL_SIGNATURE;
+  TotalFreePages   += Buffer->FreePages;
+  InsertHeadList (&mPageTablePoolList, &Buffer->NextPool);
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Walks the list to find a pool with enough free pages to allocate from.
+
+  @param[in] Pages The number of pages to allocate
+
+  @retval A pointer to the pool with enough free pages to allocate
+          from or NULL if no pool has enough free pages.
+**/
+PAGE_TABLE_POOL *
+FindPoolToAllocateFrom (
+  IN UINTN  Pages
+  )
+{
+  PAGE_TABLE_POOL  *PoolToAllocateFrom = NULL;
+  LIST_ENTRY       *CurrentEntry       = mPageTablePoolList.ForwardLink;
+
+  while (CurrentEntry != &mPageTablePoolList) {
+    PoolToAllocateFrom = CR (CurrentEntry, PAGE_TABLE_POOL, NextPool, PAGE_TABLE_POOL_SIGNATURE);
+    if (Pages <= PoolToAllocateFrom->FreePages) {
+      break;

```

Comment:
The high water mark is supposed to prevent this, but do we need an edge case condition if we don't find any pool that has enough pages?

Reply:
The edge cases are handled in AllocatePageTableMemory() when this function returns NULL